### PR TITLE
impl(rest): add AsyncRestPollingLoop

### PR DESCRIPTION
**Pull request by scotthart**
Add REST version of AsyncPollingLoop.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10576)

### CODE REVIEW

File: google/cloud/internal/async_rest_polling_loop.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+ */
+future<StatusOr<google::longrunning::Operation>> AsyncRestPollingLoop(
+    google::cloud::CompletionQueue cq,
+    future<StatusOr<google::longrunning::Operation>> op,
+    AsyncRestPollLongRunningOperation poll,
+    AsyncRestCancelLongRunningOperation cancel,
+    std::unique_ptr<PollingPolicy> polling_policy, std::string location);
+
+}  // namespace rest_internal
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
```
Comment (dbolduc) [inline google/cloud/internal/async_rest_polling_loop.h:133]: this seems out of order with the opened namespaces

File: google/cloud/internal/async_rest_polling_loop.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ *       StatusOr<Operation> op = f.get();
+ *       if (!op) return make_ready_future(op);
+ *       return AsyncPollingLoop(
+ *           std::move(cq), *std::move(op),
+ *           [stub](auto cq, auto context, auto const& r) {
+ *             return stub->AsyncGetOperation(cq, std::move(context), r);
+ *           },
+ *           [stub](auto cq, auto context, auto const& r) {
+ *             return stub->AsyncCancelOperation(cq, std::move(context), r);
+ *           },
```
Comment (dbolduc) [inline google/cloud/internal/async_rest_polling_loop.h:112]: nit, Q: is the `RestContext&` moved?

File: google/cloud/internal/async_rest_polling_loop.h [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+ *
+ * Usually the initial request to "start" the long-running operation also
+ * requires a retry loop, which is handled by this function too.
+ *
+ * Typically, library developers would use the function via
+ * `AsyncLongRunningOperation`, but as a stand-alone function it can be used in
+ * this context. First assume there is a `*Stub` class as follows:
+ *
+ * @code
+ * class BarStub {
```
Comment (dbolduc) [inline google/cloud/internal/async_rest_polling_loop.h:79]: optional nit: s/BarStub/BarRestStub/

And maybe the same thing below with `BarConnection`?

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10576)* status: 0 of 7 files reviewed, 3 unresolved discussions (waiting on @dbolduc)

___
*[`google/cloud/internal/async_rest_polling_loop.h` line 79 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10576#-NLh1ubVkw5V6vk4AFBh-r1-79:-NM-W6od0ieUFRBVxOY5:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d81d5ae777821ca6898dc58f9c17fcfaf2a53c3f/google/cloud/internal/async_rest_polling_loop.h#L79)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

optional nit: s/BarStub/BarRestStub/

And maybe the same thing below with `BarConnection`?
</blockquote></details>

Done.
___
*[`google/cloud/internal/async_rest_polling_loop.h` line 112 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10576#-NLh1ubVkw5V6vk4AFBh-r1-112:-NM-W7BfDgurOgoOWem6:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d81d5ae777821ca6898dc58f9c17fcfaf2a53c3f/google/cloud/internal/async_rest_polling_loop.h#L112)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit, Q: is the `RestContext&` moved?
</blockquote></details>

Fixed.
___
*[`google/cloud/internal/async_rest_polling_loop.h` line 133 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10576#-NLh1ubVkw5V6vk4AFBh-r1-133:-NM-W91-EY87ajmWjWMj:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d81d5ae777821ca6898dc58f9c17fcfaf2a53c3f/google/cloud/internal/async_rest_polling_loop.h#L133)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

this seems out of order with the opened namespaces
</blockquote></details>

Fixed.

**Pull request merged**